### ISSUE TITLE
KeyErrors on elements in nested refreshables

### ISSUE THREAD

**Issue by kleynjan** [zauberzeug/nicegui#1224]
### Description

Not sure if this is a bug or if I misunderstand the API. See code below. Using 1.3.5.

Steps to reproduce: 
- refresh the 'outer' card (press button 1)
- press button 2 gives KeyError in element.py line 314, trying to access self.client.elements[id] 

This nested card (# 2) is implemented as an external function. Inexplicably (to me), if the nested card is implemented as an internal function, the problem does not occur (# 3).

I would expect a refresh of the outer card to completely re-render both nested cards. Instead, it looks like the 'outer' component still uses (stale) references to the previously rendered # 2 subcomponent. 

```python
from nicegui import ui

@ui.refreshable
def nested_card_external(happy: callable) -> None:
    with ui.card().classes("outline"):
        ui.label("Nested card in external function")
        ui.button("2. External refresh function", color="deep-orange", on_click=nested_card_external.refresh)
    happy(2)

@ui.refreshable
def outer_card(happy: callable) -> None:
    @ui.refreshable
    def nested_card_internal(happy: callable) -> None:
        with ui.card().classes("outline"):
            ui.label("Nested card in internal refreshable")
            ui.button("3. Internal refresh function", on_click=nested_card_internal.refresh)
        happy(3)

    with ui.card().classes("outline"):
        ui.label("Outer card")
        ui.button("1. Outer refresh", on_click=outer_card.refresh)

        nested_card_external(happy)
        nested_card_internal(happy)
    happy(1)

@ui.page("/")
def index() -> None:
    def happy_notify(nr: int) -> None:
        ui.notify(f"Card #{nr} rendered successfully", type="positive", timeout=1500)

    outer_card(happy_notify)

ui.run()
```

Cheers, -Peter

**Comment by falkoschindler** [comment]
Thanks for bringing this up, @kleynjan!

Here is a bit more condensed version:
```py
@ui.refreshable
def a() -> None:
    ui.notify('Refreshing A')
    ui.button('Refresh A', on_click=a.refresh)

@ui.refreshable
def main() -> None:
    @ui.refreshable
    def b() -> None:
        ui.notify('Refreshing B')
        ui.button('Refresh B', on_click=b.refresh)

    ui.notify('Refreshing main')
    ui.button('Refresh main', on_click=main.refresh)
    a()
    b()

main()
```

On the one hand it somehow makes sense: When refreshing the outer container `main`, the refreshable `a` is replaced. Now the function `a` refers to something that isn't there anymore.

But on the other hand, why do we get a KeyError in
```py
descendants = [self.client.elements[id] for id in self._collect_descendant_ids()[1:]]
                   ~~~~~~~~~~~~~~~~~~~~^^^^
```

Shouldn't `_collect_descendant_ids` and `self.client.elements` be consistent? I'll have to investigate a bit more.

**Comment by falkoschindler** [comment]
Looks like this improved `prune()` function in refreshable.py fixes this issue: 
```py
    def prune(self) -> None:
        self.targets = [
            target
            for target in self.targets
            if target.container.client.id in globals.clients and target.container.id in target.container.client.elements
        ]
```

**Comment by kleynjan** [comment]
Hi Falko, you only have yourself to blame! In #1024 you show very nicely how to use decomposition into refreshables instead of complicated binds: *"I tend to simply use a piece of "refreshable" UI"* ... ;-)

Seriously, I've used that pattern in a number of places, and I've run into these key errors in other more complicated cases. Sometimes the key errors only appear when using an app with multiple clients for the first time. However, so far, I've been able to work around it.

(The lifecycle and scope of ui components is still a bit of a mystery to me. Seen from the Python side, they get created as a side effect, and after that, it's as if they live in a parallel universe...)

Cheers, -Peter
